core/node/da_dispatcher/src/da_dispatcher.rs

```diff
@@ -0,0 +1,147 @@
+use std::{future::Future, time::Duration};
+
+use anyhow::Error;
+use chrono::Utc;
+use tokio::sync::watch;
+use zksync_config::DADispatcherConfig;
+use zksync_da_layers::DataAvailabilityInterface;
+use zksync_dal::{Connection, ConnectionPool, Core, CoreDal};
+use zksync_types::L1BatchNumber;
+
+use crate::metrics::METRICS;
+
+#[derive(Debug)]
+pub struct DataAvailabilityDispatcher {
+    client: Box<dyn DataAvailabilityInterface>,
+    pool: ConnectionPool<Core>,
+    config: DADispatcherConfig,
+}
+
+impl DataAvailabilityDispatcher {
+    pub fn new(
+        pool: ConnectionPool<Core>,
+        config: DADispatcherConfig,
+        client: Box<dyn DataAvailabilityInterface>,
+    ) -> Self {
+        Self {
+            pool,
+            config,
+            client,
+        }
+    }
+
+    pub async fn run(self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let pool = self.pool.clone();
+        loop {
+            let mut conn = pool.connection_tagged("da_dispatcher").await.unwrap();
+
+            if *stop_receiver.borrow() {
+                tracing::info!("Stop signal received, da_dispatcher is shutting down");
+                break;
+            }
+
+            if let Err(err) = self.dispatch(&mut conn).await {
+                tracing::warn!("dispatch error {err:?}");
+            }
+            if let Err(err) = self.poll_for_inclusion(&mut conn).await {
+                tracing::warn!("poll_for_inclusion error {err:?}");
+            }
+
+            drop(conn);
+            tokio::time::sleep(self.config.polling_interval()).await;
+        }
+        Ok(())
+    }
+
+    async fn dispatch(&self, conn: &mut Connection<'_, Core>) -> anyhow::Result<()> {
+        let batches = conn
+            .blocks_dal()
+            .get_ready_for_da_dispatch_l1_batches(self.config.query_rows_limit() as usize)
+            .await?;
+
+        for batch in batches {
+            let dispatch_latency = METRICS.blob_dispatch_latency.start();
+            let dispatch_response = retry(self.config.max_retries(), || {
+                self.client
+                    .dispatch_blob(batch.l1_batch_number.0, batch.pubdata.clone())
+            })
+            .await
+            .map_err(Error::msg)?;
+            dispatch_latency.observe();
+
+            conn.blocks_dal()
+                .insert_l1_batch_da(batch.l1_batch_number, dispatch_response.blob_id)
+                .await?;
+
+            METRICS
+                .last_known_l1_batch
+                .set(batch.l1_batch_number.0 as usize);
+            METRICS.blob_size.observe(batch.pubdata.len());
+            tracing::info!(
+                "Dispatched a DA for batch_number: {}",
+                batch.l1_batch_number
+            );
+        }
+
+        Ok(())
+    }
+
+    async fn poll_for_inclusion(&self, conn: &mut Connection<'_, Core>) -> anyhow::Result<()> {
+        let storage_da = conn.blocks_dal().get_da_blob_awaiting_inclusion().await?;
```

Comment:
hmm so we only want to monitor the very next pending blob? Is it possible that a later blob will be included before an earlier one? I suppose it's not a big deal - we commit blobs in order anyway, just wanted to clarify. 

I suppose if it's indeed enough to only monitor one, it's better so - just for the code simplicty

Reply:
> so we only want to monitor the very next pending blob?

yes because it makes no sense for us to generate extra load on the DA layers (by polling for every dispatched blob) to cover the cases when a later blob will be included before an earlier one, while we can't proceed with committing the batch related to that later blob anyway. And yeah, it keeps the code simpler to poll for just one at a time

> maybe rename the function to 'get_first_da_blob...'

done

> nit: and btw 'StorageDataAvailability' is a strange type to be returned from this.. (if the function is named 'get_da_blob' - I'd have expected something with a 'blob' as response)

fixed that